### PR TITLE
improve error message when BUS_ADDR74 is unspecified for hard i2c/spi…

### PR DESCRIPTION
**Pull request by johnMamish**
Fix issue I talked about in [issue 478](https://github.com/YosysHQ/nextpnr/issues/478)

### CODE REVIEW

File: ice40/pack.cc [hunk 1]
```diff
@@ -1256,18 +1256,26 @@ static void pack_special(Context *ctx)
                     {std::make_tuple(id_SB_SPI, "0b0010"), Loc(25, 0, 1)},
                     {std::make_tuple(id_SB_I2C, "0b0011"), Loc(25, 31, 0)},
             };
-            if (map_ba74.find(std::make_tuple(ci->type, ci->params[ctx->id("BUS_ADDR74")].as_string())) ==
-                map_ba74.end())
-                log_error("Invalid value for BUS_ADDR74 for cell '%s' of type '%s'\n", ci->name.c_str(ctx),
-                          ci->type.c_str(ctx));
-            Loc bel_loc = map_ba74.at(std::make_tuple(ci->type, ci->params[ctx->id("BUS_ADDR74")].as_string()));
```
Comment (smunaut) [inline ice40/pack.cc:1263]: ?!? But you're removing part of the packing logic ?!?

**Comment by smunaut** [comment]
So as noted above, I don't get why your changes remove the actual packing logic.

Also you need to run your change through clang-format to get formatting consistent with project guidelines.

File: ice40/pack.cc (hunk 1 above)
Comment (johnMamish) [reply ice40/pack.cc:1263]: hahaha holy crap. I definitely meant to delete ONE line, not TWELVE.

**Comment by smunaut** [comment]
Also, it'd be nice if you could just squash and just force push 1 clean commit.
It's generally better for git bisect purposes if all the intermediates state are clean / buildable and valid.  Here it's a single logical change that doesn't really need to be split in several commits.

**Comment by johnMamish** [comment]
@smunaut Anything else I should fix to merge this?

I know this is a small change, but it would've saved me a few hours when trying to figure out why nextpnr wouldn't place a hard i2c IP block for me. Because cells_sim.v in yosys has a default parameter value for ADDR_74, I assumed that I wouldn't need to specify one myself and it would default to using the "upper left" i2c block.

**Comment by smunaut** [comment]
I mean functionally it looks goot to me.
I find it a bit weird you construct `valid_addr74_values` and then just scan through it to print the message, I don't see why you can't just combine the two iterations and never go through that intermediate step.

But in any case, I can't merge code, I'm just an interested observer because I did some work in that part of the packing code.

**Comment by johnMamish** [comment]
>I find it a bit weird you construct `valid_addr74_values` and then just scan through it to print the message
Yeah. It is a bit weird. It's to take care of the whole aesthetic issue of having a trailing comma at the end of the list of valid values. I don't write C++ much, so I don't know if stringstream easily lets me clip off the last comma. This seemed like a readable way to do it :shrug:.

**Comment by smunaut** [comment]
Huh ?

What I mean is, wouldn't this do the same thing ?

```C++
                std::stringstream ss;
                ss << "Invalid value for parameter BUS_ADDR74 for cell " << ci->name.c_str(ctx) << " of type "
                   << ci->type.c_str(ctx) << ". Valid values include: ";
                std::string delim = "";
                for (const auto &kvp : map_ba74) {
                    if (std::get<0>(kvp.first) == ci->type) {
                        ss << delim << std::get<1>(kvp.first);
                        delim = ", ";
                    }
                }
```

**Pull request closed without merging**